src/BatteryStateOfChargeService.cpp

```diff
@@ -43,21 +44,65 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return BATTERY_AMP_HOUR_CAPACITY - totalAH_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    return isCharging_;
 }
 
+
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    QTime ntime(0,0,0,0);
+    if (time.isNull())
+        return ntime;
+    else
+        return time;
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // Update your variables here.
+    if (prTime_.isNull())
+    {
+        totalAH_=BATTERY_AMP_HOUR_CAPACITY*initialStateOfChargePercent_/100;
+        prTime_=batteryData.time;
+    }
+    deltaT_=prTime_.msecsTo(batteryData.time);
+    deltaC_=(batteryData.current+preC_)/2;
+    amph_=deltaT_*deltaC_/3600000;
```

Comment:
Make sure you define any constants that you use for calculations (like the 3600000 here). Using a constant with a descriptive name instead of a "magic number" will help make the code easier to understand and more maintainable.